Lockable storage units, WIP

drcrazy opened the pull request.

drcrazy commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -54,11 +58,35 @@
 public class StorageUnitBaseBlockEntity extends MachineBaseBlockEntity
 	implements InventoryProvider, IToolDrop, IListInfoProvider, IContainerProvider {
 
+	private String logPref() {
```
This might go to common log functionality

drcrazy commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -222,6 +265,11 @@ public ItemStack processInput(ItemStack inputStack) {
 
 	public boolean isSameType(ItemStack inputStack) {
+		if (getLocked()) {
+			System.out.printf("+++ SUB#isSameType: locked: [%s]\n", lockedItem);
```
Just log with debug log level

Sturmlilie replied:
Sorry, I left all this stuff in because it's not meant to go into a formal PR.

drcrazy replied:
It is absolutely fine to have extended logging. Might help with debugging in future. Just log with debug level

Sturmlilie committed:
StorageUnitBlock: Implement locking mechanism

A SU that is non-empty can be "locked-in" into the contained
item type. Once locked, the SU will continue behaving as if
that item type is contained within, even when the actual count
drops to zero.

This makes it easier to designate specific SUs to only one item.

Sturmlilie commented:
@drcrazy If you want generic debug log messages, you're going to have to tell me what specifically you want logged. My earlier logging was just scaffolding for learning how the codebase works and getting this implemented.

Sturmlilie commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -54,11 +58,15 @@
+	// NBT keys
+	private static final String TAG_LOCKED_ITEM = "lockedItem";
+
+	private static final Item NULL_ITEM = Items.AIR;
```
I am unsure whether this isn't causing a race with the static initializer of `Items.AIR`.

modmuss50 replied:
I would use an itemstack here, as it should lock on the NBT value as well.

modmuss50 commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -54,11 +58,15 @@
 	private static final int OUTPUT_SLOT = 1;
 
+	// NBT keys
+	private static final String TAG_LOCKED_ITEM = "lockedItem";
```
Not sure this is needed here tbh, not the end of the world.

modmuss50 commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -68,6 +76,10 @@
 
+	// A locked storage unit will continue behaving as if it contains
+	// the locked-in item, even if the stored amount drops to zero.
+	private Item lockedItem = NULL_ITEM;
```
Again, use an itemstack, and set it to empty.

modmuss50 commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -222,6 +244,10 @@ public ItemStack processInput(ItemStack inputStack) {
 
 	public boolean isSameType(ItemStack inputStack) {
+		if (isLocked()) {
+			return inputStack.getItem() == lockedItem;
```
When replaced with a stack use a similar check as bellow.

modmuss50 commented on src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:
```diff
@@ -377,13 +413,50 @@ public void onBreak(World world, PlayerEntity playerEntity, BlockPos blockPos, B
 		// Inventory gets dropped automatically
 	}
 
+	// The int methods are only for ContainerBuilder.sync()
```
You should be able to sync booleans?

modmuss50 replied:
Even better would be sycning the locked type so the user can see what type of item the container has been locked to.

Sturmlilie replied:
I have no idea how this `sync()` business works or why it's needed, I just copied it from the two other block entities that also have a lock button in their GUI (autocraft table and rolling machine). Will have to look into it.

modmuss50 commented on src/main/java/techreborn/client/gui/GuiStorageUnit.java:
```diff
@@ -29,9 +29,14 @@
 
 public class GuiStorageUnit extends GuiBase<BuiltContainer> {
 
+	private static final int LOCK_BUTTON_X = 145;
+	private static final int LOCK_BUTTON_Y = 4;
```
Not sure we need these here, id inline them.

modmuss50 reviewed (changes requested):
Looks like a good start, I would move to using ItemStack as the locking "key" would allow for NBT checking and simpler code.

Sturmlilie commented:
Changes since last version:
* Switch to ItemStack for NBT support
* Un-DRY-ify
* Push lock button further to the right to accomodate longer strings

I tried understanding the `sync()` stuff I copied from `RollingMachine` but failed to grasp it; in `ContainerBlockEntityInventoryBuilder` there is a comment about automatic syncing of the value via a custom packet, but I couldn't find where this packet exchange happens. Also if it's automatically synchronizing, why do I need to manually send / handle a packet still?

Sturmlilie commented:
Removed superfluous imports.
